### DISCUSSION TITLE
How to make an agoric address using cosmjs?

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#5830] [Q&A] [answered]
@dtribble asked for a quick example of making an address using the [cosmjs](https://github.com/cosmos/cosmjs) library.

**Comment by dckc** [comment]
The basic outline is:

```js
import { Random } from '@cosmjs/crypto';
import { DirectSecp256k1Wallet } from '@cosmjs/proto-signing';

export const bech32Config = {
  bech32PrefixAccAddr: 'agoric',
 // ...
};

const makeAddr = async () => {
    const seed = Random.getBytes(32);
    const wallet = await DirectSecp256k1Wallet.fromKey(seed, bech32Config.bech32PrefixAccAddr);
    const accounts = await wallet.getAccounts()
    console.log({ accounts });
};

makeAddr().catch(err => console.error(err));
```

and out comes
```
{
  accounts: [
    {
      address: 'agoric1pnhkj4ay6qt60pk6t8efmlf5775s8z9a9s0v97',
...
```

for a complete example including `package.json` and all that, see https://gist.github.com/dckc/7aa425d6986dd714a76334ff8c0bb9e5

cc @dtribble

**Reply by 0xpatrickdev** [reply, marked as answer]
Hey @dckc, I was looking into this more after reading https://github.com/Agoric/agoric-sdk/discussions/6994. 

In order to match an address generated by the Keplr client, I found that I needed to use `coinType: 564`. 

Looking [here](https://github.com/cosmos/chain-registry/blob/master/agoric/chain.json#L12), Agoric's slip44 parameter is 564. In the [keplr registry](https://github.com/chainapsis/keplr-chain-registry/blob/main/cosmos/agoric.json#L16), I see `bip44.coinType` = 564, but further down I also see an `alternativeBIP44s` field which contains 118. **_This leads me to believe both 118 and 564 are valid, but I wanted to see if you could provide more color here._**

@tgrecojs also shared this doc which was a helpful resource for me: https://github.com/confio/cosmos-hd-key-derivation-spec/blob/main/README.md

Here is an implementation with `coinType: 564` - `DirectSecp256k1Wallet` does not allow us to specify coinType/hdPath (it uses the ATOM default 118), so below `DirectSecp256k1HdWallet` is used.

```js
import { Random, Bip39, stringToPath, Slip10RawIndex } from "@cosmjs/crypto";
import { DirectSecp256k1HdWallet } from "@cosmjs/proto-signing";

export const bip44Config = {
  coinType: 564,
};

export const bech32Config = {
  bech32PrefixAccAddr: "agoric",
  // ...,
};

const getAgoricHdPath = (coinType = bip44Config.coinType, account = 0) =>
  stringToPath(`m/44'/${coinType}'/${account}'/0/0`);

// alternative to stringToPath from cosmjs
const getAgoricHdPathWithSlip = (
  coinType = bip44Config.coinType,
  account = 0
) => [
  Slip10RawIndex.hardened(44),
  Slip10RawIndex.hardened(coinType),
  Slip10RawIndex.hardened(account),
  Slip10RawIndex.normal(0),
  Slip10RawIndex.normal(0),
];

const makeAddr = async () => {
  const seed = Random.getBytes(32);
  const mnemonic = Bip39.encode(seed); // fromKey() not supported, convert to mnemonic first
  const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic.data, {
    prefix: "agoric",
    hdPaths: [getAgoricHdPath()],
  });
  const accounts = await wallet.getAccounts();
  return { seed, mnemonic, accounts };
};

const addressFromMnemonic = async (mnemonic) => {
  const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {
    prefix: bech32Config.bech32PrefixAccAddr,
    hdPaths: [getAgoricHdPath()],
  });
  const accounts = await wallet.getAccounts();
  return accounts;
};
```

**Reply by 0xpatrickdev** [reply]
Also found [`DirectSecp256k1HdWallet.generate()`](https://github.com/cosmos/cosmjs/blob/33271bc51cdc865cadb647a1b7ab55d873637f39/packages/proto-signing/src/directsecp256k1hdwallet.ts#L152-L166) which is an abstraction of the `Random` and `Bip39` logic.

```js
import { stringToPath } from "@cosmjs/crypto";
import { DirectSecp256k1HdWallet } from "@cosmjs/proto-signing";

const getAgoricHdPath = (coinType = 564, account = 0) =>
  stringToPath(`m/44'/${coinType}'/${account}'/0/0`);

/**
 * @param length The number of words in the mnemonic (12, 15, 18, 21 or 24).
 */
const makeAddr = async (length = 24) => {
  const wallet = await DirectSecp256k1HdWallet.generate(length, {
    prefix: "agoric",
    hdPaths: [getAgoricHdPath()],
  });
  const accounts = await wallet.getAccounts();
  return { mnemonic: wallet.mnemonic, seed: wallet.seed, accounts };
};

makeAddr().then(console.log).catch(console.error);
```

**Reply by dckc** [reply]
> Looking [here](https://github.com/cosmos/chain-registry/blob/master/agoric/chain.json#L12), Agoric's slip44 parameter is 564. In the [keplr registry](https://github.com/chainapsis/keplr-chain-registry/blob/main/cosmos/agoric.json#L16), I see `bip44.coinType` = 564, but further down I also see an `alternativeBIP44s` field which contains 118. **_This leads me to believe both 118 and 564 are valid, but I wanted to see if you could provide more color here._**

My rough understanding is that 564 is a coin type we registered, but it's necessary to use 118, the default cosmos coinType, in certain interop situations; in particular, when using a ledger.

It seems to be defined in `config.go`:

https://github.com/Agoric/agoric-sdk/blob/8ccc9cc746c35d6e476bf5de8570f7ce39060a13/golang/cosmos/agoric/config.go#L12-L13

`git blame` leads to a rename/refactor; the earliest I can trace it back to is 2021-12-09 29535ded8 but I thought I had seen it before that.

@michaelfig did I miss anything?

**Reply by 0xpatrickdev** [reply]
Ok cool, thanks Dan. I tested with a ledger and am seeing that the 118 derivation path is used to generate the address. When imported to Keplr (via "Connect Ledger"), an additional label is added that shows the 118 derivation path is used.
<img width="240" alt="image" src="https://user-images.githubusercontent.com/11021913/220525815-8a7efb40-342e-4707-a00f-bfdc41a726eb.png"> 

One more question has since surfaced - I am trying to learn more about the `account` parameter. [This doc](https://github.com/confio/cosmos-hd-key-derivation-spec/blob/main/README.md) is leading me to believe that account should always be `0` for Cosmos chains, and `addressIndex` is what we should be incrementing to get additional accounts. The `getAgoricHdPath ` function would instead look like: 

```js
const getAgoricHdPath = (coinType = 564, addressIndex = 0) =>
  stringToPath(`m/44'/${coinType}'/0'/0/${addressIndex}`);
```


I found a conflicting implementation here, so perhaps this understanding is incorrect. Maybe @kriskowal has some insights. 


https://github.com/Agoric/agoric-sdk/blob/3ff8f54a4eadf771580be2a36134294b716fbe36/packages/agoric-cli/src/publish.js#L38-L39

https://github.com/Agoric/agoric-sdk/blob/3ff8f54a4eadf771580be2a36134294b716fbe36/packages/agoric-cli/src/publish.js#L263-L266
